core/drivers/crypto/caam/acipher/caam_rsa.c

```diff
@@ -618,29 +649,39 @@ static TEE_Result do_gen_keypair(struct rsa_keypair *key, size_t key_size)
 	caam_desc_add_word(desc, DESC_HEADER(0));
 
 	caam_desc_add_word(desc, 0);
-	caam_desc_add_word(desc, PDB_RSA_KEY_P_SIZE(genkey.p.length));
+	caam_desc_add_word(desc, PDB_RSA_KEY_P_SIZE(genkey.p.sec_size));
 	caam_desc_add_word(desc, PDB_RSA_KEY_N_SIZE(genkey.n.length) |
 					 PDB_RSA_KEY_E_SIZE(genkey.e.length));
 
-	caam_desc_add_ptr(desc, genkey.p.paddr);
-	caam_desc_add_ptr(desc, genkey.q.paddr);
+	caam_desc_add_ptr(desc, genkey.p.buf.paddr);
+	caam_desc_add_ptr(desc, genkey.q.buf.paddr);
 	caam_desc_add_ptr(desc, genkey.e.paddr);
 	caam_desc_add_ptr(desc, genkey.n.paddr);
-	caam_desc_add_ptr(desc, genkey.d.paddr + sizeof(uint32_t));
-	caam_desc_add_ptr(desc, genkey.d.paddr);
+	caam_desc_add_ptr(desc, genkey.d.buf.paddr);
+	caam_desc_add_ptr(desc, virt_to_phys(size_d_gen_val_ptr));
 
 	if (genkey.format > RSA_PRIVATE_KEY_FORMAT_2) {
-		caam_desc_add_ptr(desc, genkey.dp.paddr);
-		caam_desc_add_ptr(desc, genkey.dq.paddr);
-		caam_desc_add_ptr(desc, genkey.qp.paddr);
-		caam_desc_add_word(desc, RSA_FINAL_KEY(ALL));
-
-		cache_operation(TEE_CACHEFLUSH, genkey.dp.data,
-				genkey.dp.length + genkey.dq.length +
-					genkey.qp.length);
-
+		caam_desc_add_ptr(desc, genkey.dp.buf.paddr);
+		caam_desc_add_ptr(desc, genkey.dq.buf.paddr);
+		caam_desc_add_ptr(desc, genkey.qp.buf.paddr);
+
+		if (key_type == CAAM_KEY_PLAIN_TEXT)
+			caam_desc_add_word(desc, RSA_FINAL_KEY(ALL, NONE));
+		else if (key_type == CAAM_KEY_BLACK_ECB)
+			caam_desc_add_word(desc, RSA_FINAL_KEY(ALL, ECB));
+		else if (key_type == CAAM_KEY_BLACK_CCM)
```

Comment:
Add no word, if `key_type` isn't `CAAM_KEY_BLACK_CCM` either?

Same below.

Reply:
`key_type` have three valid values. This `if .. else if` covers all valid values.